### ISSUE TITLE
Dependency violation when deleting a Subnet: ENI still attached to ELB 

### ISSUE THREAD

**Issue by hashibot** [hashicorp/terraform-provider-aws#151]
_This issue was originally opened by @rokka-n as hashicorp/terraform#6318. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi,

I'm currently seeing terraform tries to destroy a subnet that was used in ELB and fails. I think its because  ENI interface that belongs to the subnet is still attached to ELB. The subnet itself is removed from ELB correctly via ASG routine.     
### Terraform Version

v0.6.14
### Actual Behavior

When I remove a subnet from vars, terraform correctly removes it from ASG/ELB. But it hangs at the subnet destroying step.

Workaround is to manually detach ENI(s) that belong to subnet and delete it.

```
aws_autoscaling_group.wp_asg: Refreshing state... (ID: wp-staging-qmlts6tb7fb3va6r3xe5glsie4-asg)
aws_subnet.blog_subnets.2: Refreshing state... (ID: subnet-c21543ff)
aws_subnet.blog_subnets.2: Destroying...

aws_subnet.blog_subnets.2: Destruction complete

Apply complete! Resources: 0 added, 0 changed, 1 destroyed.
```
### Steps to Reproduce

Create an ELB with ASG that uses few subnets. Remove 1 subnet. 
Watch ASG/ELB gets updated properly, but terraform hangs on subnet termination. If I try to delete subnet manually in aws interface - it complaines about attached ENI.   

```
resource "aws_elb" "wp_elb" {
name = foo
subnets = ["${split(",", join(",", aws_subnet.app_subnets.*.id))}"]
}
```

Could be related to hashicorp/terraform#934

**Comment by radeksimko** [comment]
Hi @rokka-n
just to clarify - this is not a dependency problem, but rather eventual consistency or generally a problem with undocumented actions that happen behind the scenes in AWS. Some services manage their own ENIs but do so asynchronously with delays significant enough to cause problems for any tool managing the rest of the infrastructure.

https://github.com/terraform-providers/terraform-provider-aws/pull/1036 was merged a while back to address very similar problem - detaching ENIs on ELB deletion. It probably wouldn't be too difficult to crib something from that PR and use it in this context. I might take a look at it at some point.

**Comment by rokka-n** [comment]
Thanks for looking into this, I'd like to close this issues since I haven't seen this happened for some time.

Probably just weird edge case.

**Comment by awilkins** [comment]
Not just a weird edge case, this happens reliably for us with any plan that includes a delete/recreate of a subnet attached to an ELB.

I would read it as a bug determining the dependency graph because the subnet identifier is referenced directly in the ELB concerned. We even tried explicitly declaring  a `depends_on` relationship from the ELB to the subnet - it still tries to delete the subnet without first deleting the ELB or detaching the subnet from it. Is there some kind of special case in the code that treats ELB -> subnet dependencies as "soft"?

**Comment by radeksimko** [comment]
I managed to reliably reproduce this using the following config:
https://gist.github.com/radeksimko/4132453a6cf6edc3ccb61c8bbe31e076

As @awilkins mentioned already this is triggered by deletion of a subnet which is associated with an LB.

Sadly though there doesn't seem to be anything we can realistically do from Terraform's perspective. Even if we got the ordering of operations right and managed to disassociate the subnet from LB before attempting to delete it, Amazon wouldn't let us detach these ENIs as they hold them until the LB is fully destroyed. You will get the following error when attempting to do so:

> An error occurred (AuthFailure) when calling the DetachNetworkInterface operation: You do not have permission to access the specified resource.

I am not sure if this is by design. Perhaps we could get some explanation from the LB/EC2 teams?
